src/CMakeLists.txt

```diff
@@ -0,0 +1,150 @@
+# Must be included before CMAKE_INSTALL_INCLUDEDIR is used.
+include(GNUInstallDirs)
+set(${PROJECT_NAME}_installables "")
+
+add_library(${PROJECT_NAME} SHARED "")
+target_sources(${PROJECT_NAME} PRIVATE
+  precomputed_ecmult.c
+  precomputed_ecmult_gen.c
+  secp256k1.c
+)
+target_include_directories(${PROJECT_NAME}
+  INTERFACE
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
+)
+if(WIN32)
+  target_compile_definitions(${PROJECT_NAME} PRIVATE DLL_EXPORT)
+endif()
+set_target_properties(${PROJECT_NAME} PROPERTIES
+  VERSION "${${PROJECT_NAME}_LIB_VERSION_CURRENT}.${${PROJECT_NAME}_LIB_VERSION_AGE}.${${PROJECT_NAME}_LIB_VERSION_REVISION}"
+  SOVERSION "${${PROJECT_NAME}_LIB_VERSION_CURRENT}"
+)
+if(SECP256K1_BUILD_SHARED)
+  list(APPEND ${PROJECT_NAME}_installables ${PROJECT_NAME})
+else()
+  set_target_properties(${PROJECT_NAME} PROPERTIES
+    EXCLUDE_FROM_ALL TRUE
+  )
+endif()
+
+add_library(${PROJECT_NAME}_static STATIC "")
+target_sources(${PROJECT_NAME}_static PRIVATE
+  precomputed_ecmult.c
+  precomputed_ecmult_gen.c
+  secp256k1.c
+)
+target_include_directories(${PROJECT_NAME}_static
+  INTERFACE
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
+)
+if(NOT MSVC)
+  set_target_properties(${PROJECT_NAME}_static PROPERTIES
+    OUTPUT_NAME ${PROJECT_NAME}
+  )
+endif()
+if(SECP256K1_BUILD_STATIC)
+  list(APPEND ${PROJECT_NAME}_installables ${PROJECT_NAME}_static)
+else()
+  set_target_properties(${PROJECT_NAME}_static PROPERTIES
+    EXCLUDE_FROM_ALL TRUE
+  )
+endif()
+
+add_library(binary_interface INTERFACE)
+target_compile_definitions(binary_interface INTERFACE
+  $<$<C_COMPILER_ID:MSVC>:_CRT_SECURE_NO_WARNINGS>
+)
+
+add_library(link_shared INTERFACE)
+target_link_libraries(link_shared INTERFACE
+  ${PROJECT_NAME}
+)
+
+add_library(link_static INTERFACE)
+target_link_libraries(link_static INTERFACE
+  ${PROJECT_NAME}_static
+)
+
+if(SECP256K1_BUILD_BENCHMARK)
+  add_executable(bench bench.c)
+  target_link_libraries(bench binary_interface link_shared)
```

Comment:
> it'll always link bench with the shared library, even when the shared library isn't being built. It should just link with whatever is being built - probably shared if shared is available, and static otherwise.

Going to implement this suggestion.

> And below the opposite problem occurs: it always links with the static library, even when the static library isn't being built. In fact, it shouldn't be linked with either the static or shared library; it just needs to link the common object files (precomputed stuff, asm if included).

I believe there is no difference between linking against object files and linking against a static library, which is an archive of object files, no?

Reply:
Easy to reproduce with:
```
$ gcc src/bench_internal.c src/precomputed_ecmult.c src/precomputed_ecmult_gen.c src/secp256k1.c
```

Reply:
Thanks! [Updated](https://github.com/bitcoin-core/secp256k1/pull/1113#issuecomment-1396093991).